### PR TITLE
Add support for creating tiles when adding multiple atlases to a tileset

### PR DESCRIPTION
**Pull request by jamesmintram**
Adds support for "batch" creating tiles when dragging multiple atlases into a tileset. See: https://github.com/godotengine/godot/issues/69273

### CODE REVIEW

File: editor/plugins/tiles/tile_set_atlas_source_editor.cpp [hunk 1]
```diff
@@ -2195,55 +2195,70 @@ void TileSetAtlasSourceEditor::edit(Ref<TileSet> p_tile_set, TileSetAtlasSource
...
-						for (int region_y = region.get_position().y; region_y < region.get_end().y; region_y++) {
-							if (texture->is_pixel_opaque(region_x, region_y)) {
-								is_opaque = true;
+void TileSetAtlasSourceEditor::_auto_create_tiles_cancel() {
+	hackBodge.clear();
+}
+
+void TileSetAtlasSourceEditor::_auto_create_tiles(Vector<Ref<TileSetAtlasSource>> const &new_tile_set_atlas_sources) {
+
+	Ref<EditorUndoRedoManager> &undo_redo = EditorNode::get_undo_redo();
```
Comment (jamesmintram) [inline editor/plugins/tiles/tile_set_atlas_source_editor.cpp:2215]: Single undo scope for all tile generation

File: editor/plugins/tiles/tile_set_atlas_source_editor.cpp [hunk 2]
```diff
@@ -2195,55 +2195,70 @@ void TileSetAtlasSourceEditor::edit(Ref<TileSet> p_tile_set, TileSetAtlasSource
 	_update_current_tile_data_editor();
 }
 
-void TileSetAtlasSourceEditor::init_source() {
+void TileSetAtlasSourceEditor::init_source(Vector<Ref<TileSetAtlasSource>> newSources) {
+	hackBodge = newSources;
```
Comment (jamesmintram) [inline editor/plugins/tiles/tile_set_atlas_source_editor.cpp:2199]: `hackBodge` is used to hold the new sources while waiting for the user to close the dialog.

File: editor/plugins/tiles/tile_set_atlas_source_editor.h [hunk 3]
```diff
@@ -259,7 +259,9 @@ class TileSetAtlasSourceEditor : public HSplitContainer {
 	void _unhandled_key_input(const Ref<InputEvent> &p_event);
 
 	// -- Misc --
+	void _auto_create_tiles(Vector<Ref<TileSetAtlasSource>> const &new_tile_set_atlas_sources);
```
Comment (jamesmintram) [inline editor/plugins/tiles/tile_set_atlas_source_editor.h:262]: Better naming?

File: editor/plugins/tiles/tile_set_atlas_source_editor.cpp (hunk 2 above)
Comment (jamesmintram) [reply editor/plugins/tiles/tile_set_atlas_source_editor.cpp:2199]: Renamed to `pendingAtlases`

**Comment by jamesmintram** [comment]
Video of the changes in action, note the Undo error that appears is a pre-existing issue and (probably) out of scope for this change:

batch-add-tileset-atlas

File: editor/plugins/tiles/tile_set_atlas_source_editor.h [hunk 4]
```diff
@@ -119,6 +119,8 @@ class TileSetAtlasSourceEditor : public HSplitContainer {
 
 	bool tile_set_changed_needs_update = false;
 
+	Vector<Ref<TileSetAtlasSource>> pendingAtlases;
```
Comment (akien-mga) [inline editor/plugins/tiles/tile_set_atlas_source_editor.h:122]: Please use snake_case for identifiers. You might see from surrounding code that we never use camelCase ;)

File: editor/plugins/tiles/tile_set_atlas_source_editor.h (hunk 3 above)
Comment (akien-mga) [reply editor/plugins/tiles/tile_set_atlas_source_editor.h:264]: Should be `const Vector<...> &`, and the parameter should start with `p_`.

File: editor/plugins/tiles/tile_set_atlas_source_editor.h [hunk 5]
```diff
@@ -275,7 +279,7 @@ class TileSetAtlasSourceEditor : public HSplitContainer {
 
 public:
 	void edit(Ref<TileSet> p_tile_set, TileSetAtlasSource *p_tile_set_source, int p_source_id);
-	void init_source();
+	void init_sources(Vector<Ref<TileSetAtlasSource>> newSources);
```
Comment (akien-mga) [inline editor/plugins/tiles/tile_set_atlas_source_editor.h:282]:
```suggestion
	void init_sources(Vector<Ref<TileSetAtlasSource>> p_new_sources);
```

**Comment by groud** [comment]
I think the idea is fine. The implementation looks correct enough, but as the popup and automatic creation of tiles would not depends anymore on the currently edited atlas source, the logic to create those tiles should be moved outside of `TileSetAtlasSourceEditor`. Could you move all this code to `TileSetEditor` instead ?

**Comment by coppolaemilio** [comment]
Hello @jamesmintram, thanks for making the PR! Would you be up to make the latest requested changes or do you want someone else to take over? 

---

Fixes https://github.com/godotengine/godot/issues/69273

**Comment by akien-mga** [comment]
Superseded by #79678.

**Pull request closed without merging**